pubtools/pulplib/_impl/client/repo_lock.py

```diff
@@ -0,0 +1,161 @@
+        )
+        self.expires = (
+            datetime.fromtimestamp(kwargs["expires"])
+            if "expires" in kwargs
+            else self.created + timedelta(seconds=LOCK_EXPIRATION_OFFSET)
+        )
+
+    @property
+    def as_json(self):
+        data = {
+            "id": self.id,
+            "created": datetime.timestamp(self.created),
+            "valid_from": datetime.timestamp(self.valid_from),
+            "expires": datetime.timestamp(self.expires),
+        }
+        return json.dumps(data)
+
+    def __lt__(self, other):
+        return self.created < other.created
+
+    def __eq__(self, other):
+        return self.id == other.id
+
+    @staticmethod
+    def from_json_data(json_data):
+        return LockClaim(**json_data)
+
+    @property
+    def is_valid(self):
+        return self.expires > datetime.now()
+
+
+class RepoLock(object):
+    def __init__(self, repo_id, request_func):
+        self._repo_id = repo_id
+        self._request_func = request_func
+        self._lock_claim = None
+
+    def __enter__(self):
+        self.remove_expired_locks()
+        self._lock_claim = LockClaim()
+        self.submit_lock_claim()
+        self.await_lock_activation()
+
+    def __exit__(self, exc_type, exc_value, tb):
+        self.delete_lock_claim()
+        self._lock_claim = None
+
+    def get_repo_lock_data(self, filter_invalid=False):
+        print("Getting current repo lock data")
+        repo_url = REPO_ENDPOINT_BASE % self._repo_id
+
+        repo_raw_f = self._request_func(repo_url, method="GET")
+
+        notes_f = f_map(repo_raw_f, lambda data: (data.get("notes") or {}))
+        locks_f = f_map(
+            notes_f,
+            lambda notes: {key: notes[key] for key in notes if LOCK_CLAIM_STR in key},
+        )
+
+        locks_f = f_map(
+            locks_f,
+            lambda locks: [
+                LockClaim.from_json_data(json.loads(l)) for l in locks.values()
+            ],
+        )
+        parsed_locks = locks_f.result()
+        parsed_locks.sort()
+        if filter_invalid:
+            parsed_locks = [l for l in parsed_locks if l.is_valid]
+        return parsed_locks
+
+    def await_lock_activation(self):
+        sleep(max((self._lock_claim.valid_from - datetime.now()).seconds, 0))
+        while True:
+            lock_claims = self.get_repo_lock_data(filter_invalid=True)
+            # If our claim is the oldest valid lock claim.
+            if self._lock_claim == lock_claims[0]:
+                print("Lock is now valid!")
+                return
+            else:
+                try:
+                    print(
+                        "Lock not yet valid: position in queue "
+                        + str(lock_claims.index(self._lock_claim))
+                    )
+                    sleep(
+                        AWAIT_LOCK_ACTIVATION_SLEEP
+                        * lock_claims.index(self._lock_claim)
+                    )
+                except ValueError:
```

Comment:
I'm not sure that we really need to catch this at all.  This would happen if, for example, the lock notes contain corrupt data?

You can't really do anything other than raise an exception in such cases, and if you raise something other than *the original* exception it arguably just obscures the error.  Also, if the backtrace shows more than one exception, a lot of developers will not read it properly and will randomly pick whichever of the exceptions they noticed first as being the root cause...